### PR TITLE
Lock Contract MVP0

### PR DESCRIPTION
**Pull request by faustovanin**
## Summary by CodeRabbit

- **New Features**
	- Introduced `ERC721GenericMetadata` contract for dynamic NFT attribute management.
	- Added `GenericTypedMessage` contract for secure verification of signed messages.
	- Implemented `RecyCertificate` contract with upgradeable functionality for NFTs.
	- Enhanced `TimeLock` contract with new certificate authorization features.
	- Added utility for generating EIP-712 compliant domain separators.

- **Bug Fixes**
	- Improved security and usability in the locking mechanism of the `TimeLock` contract.

- **Chores**
	- Updated dependencies in `package.json` to include new libraries and ensure compatibility.

### CODE REVIEW

File: contracts/ERC721GenericMetadata.sol [hunk 1]
```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.16;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts/utils/Strings.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlEnumerableUpgradeable.sol";
+import "./utils/ModifiedEnumerableMapUpgradeable.sol";
+// import "./utils/Roles.sol";
```
Comment (afmeirelles) [inline contracts/ERC721GenericMetadata.sol:8]: Can be deleted

File: contracts/utils/Roles.sol [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: MIT
```
Comment (afmeirelles) [inline contracts/utils/Roles.sol:1]: Don't think this contract is being used

**Review by afmeirelles — CHANGES_REQUESTED** [review]

File: contracts/TimeLock.sol [hunk 3]
```diff
@@ -60,41 +85,113 @@ contract TimeLock is
...
     }
 
     /**
      * @notice Lock cRECY tokens for a period of time
      * @param amount token amount to lock
      */
-    function lock(uint256 amount) external whenNotPaused nonReentrant {
-        _registerLocker(_msgSender());
+    function lock(
+        uint256 amount,
```
[the remark this answers is not in the document]
Comment (faustovanin) [reply contracts/TimeLock.sol:134]: No, tons represent the recycling total weight and amount the token to be locked.

**Review by afmeirelles — APPROVED** [review]
LGTM

**Review by yurimutti — APPROVED** [review]

**Pull request merged**